src/libraries/System.IO.FileSystem/tests/RandomAccess/NoBuffering.Windows.cs

```diff
@@ -180,6 +180,37 @@ public async Task WriteAsyncUsingMultipleBuffers(bool asyncOperation, bool async
             Assert.Equal(content, File.ReadAllBytes(filePath));
         }
 
+        [Fact]
+        public async Task ReadWriteAsyncUsingNonPageSizedMultipleBuffers()
+        {
+            string filePath = GetTestFilePath();
+            // The Windows scatter/gather APIs accept segments that are exactly one page long.
+            // Combined with the FILE_FLAG_NO_BUFFERING's requirements, the segments must also
+            // be aligned at page size boundaries and have a size of a multiple of the page size.
+            // Using segments with a length of twice the page size adheres to the second requirement
+            // but not the first. The RandomAccess implementation will see it and issue sequential
+            // read/write syscalls per segment, instead of one scatter/gather syscall.
+            // This test verifies that fallback behavior.
+            int bufferSize = Environment.SystemPageSize * 2;
+            int fileSize = bufferSize * 2;
+            byte[] content = RandomNumberGenerator.GetBytes(fileSize);
+
+            using (SafeFileHandle handle = File.OpenHandle(filePath, FileMode.CreateNew, FileAccess.ReadWrite, FileShare.None, FileOptions.Asynchronous | NoBuffering))
+            using (SectorAlignedMemory<byte> buffer = SectorAlignedMemory<byte>.Allocate(fileSize))
+            {
+                Memory<byte> firstHalf = buffer.Memory.Slice(0, bufferSize);
+                Memory<byte> secondHalf = buffer.Memory.Slice(bufferSize);
+
+                content.AsSpan().CopyTo(buffer.GetSpan());
+                await RandomAccess.WriteAsync(handle, new ReadOnlyMemory<byte>[] { firstHalf, secondHalf }, 0);
```

Comment:
I'm confused what this is testing.  The comment suggests this doesn't meet the "aligned at page size boundaries" requirement.  What here guarantees that isn't the case?

Reply:
@stephentoub  `ReadFileScatter` and `WriteFileGather` requirement is that each buffer must be of `SystemPageSize` size. It can't be a multiple of it, it must be equal. In this test the `bufferSize = Environment.SystemPageSize * 2` and hence the requirement is not met.

Reply:
I guess my comment was not clear enough. The "first requirement" is the "segments [...] are exactly one page long" on the first sentence.

The second requirement is the "aligned at page size boundaries" that you thought it was the first, and the third requirement is the "have a size of a multiple of the page size", which is superseded by the first.

This test passes two buffers that are aligned but are twice the page size long. It meets the requirements of `FILE_FLAG_NO_BUFFERING` (the second and the third), but not the first requirement (of Windows scatter/gather APIs). The `RandomAccess methods will detect it and use multiple syscalls.

Reply:
Thanks.  Can we fix the comment? Also,  what tests are validating the other requirements / fallback in those cases?

Reply:
The comment was fixed. Further testing of scatter/gather I/O could follow the template of the test I provided but is outside this PR's scope.

Reply:
> outside this PR's scope

Is it? I want to make sure I understand.  Isn't this PR causing fallback to sequential on multiple conditions but this test is only validating behavior for one of those conditions? Or have I misunderstood?

Reply:
We open the file with `FILE_FLAG_NO_BUFFERING`. The only way for an I/O operation to fulfill that flag's requirements but not the scatter/gather APIs' requirements is if the buffers are aligned and their size is a multiple of the page size, but not exactly it.

If execution reaches https://github.com/dotnet/runtime/pull/57424/files#diff-8e1fc498b4c506832ef1f17dc9eec9324657413163179b809f45b82830307182R476, the I/O operation is doomed to fail. We have previously checked that we are on unbuffered mode and we are going to read/write from/to an unaligned memory location.

I thought of removing this alignment check since it does not prevent code from failing but I am not sure whether it is a good idea.

Reply:
> We have previously checked that we are on unbuffered mode and we are going to read/write from/to an unaligned memory location.

I see us checking NoBuffering:
https://github.com/dotnet/runtime/blame/a3b83b91920c88041649ec229cbc9ece446cd5a6/src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Windows.cs#L424
Where are we checking alignment prior to your PR?

Reply:
Nowhere.

Another thing I noticed is that `FILE_FLAG_NO_BUFFERING` [requires the buffers to be aligned in _disk sector_ sizes](https://docs.microsoft.com/en-us/windows/win32/fileio/file-buffering#alignment-and-file-access-requirements), while the scatter/gather APIs require the buffers to be aligned in _memory page_ sizes.

Can the memory page size ever be larger than the disk sector size?

Reply:
>>>>> Further testing of scatter/gather I/O could follow the template of the test I provided but is outside this PR's scope.

>>>> Isn't this PR causing fallback to sequential on multiple conditions but this test is only validating behavior for one of those conditions?

>>> We have previously checked that we are on unbuffered mode and we are going to read/write from/to an unaligned memory location.

>> Where are we checking alignment prior to your PR?

> Nowhere.

Ok, so just to try to restate my comment:
The PR added new checks / new code paths, but only added a test for a subset of them.  I was suggesting the PR should have included tests for all of them. (Regardless, I appreciate your efforts to create the PR in the first place.)

> Can the memory page size ever be larger than the disk sector size?

Sure.  The machine I'm using has disks with 512 bytes / sector and 4K memory pages.

Reply:
> We have previously checked that we are on unbuffered mode and we are going to read/write from/to an unaligned memory location.

I think we have misunderstood the meaning of "previously" here. You thought I meant "before the PR" but I mean "earlier in code".

> Can the memory page size ever be larger than the disk sector size?

What I wanted to ask is whether the "disk sector aligned" requirement is stronger or weaker than the "memory page aligned". After some googling, it looks like it is weaker on common cases, but to be robust against extreme cases (like future hard drives where the sector size is bigger than 4K), the memory page alignment check will stay. I will reword the comments to reflect the difference between disk sector and memory page size.

---

> The PR added new checks / new code paths, but only added a test for a subset of them.

Let's enumerate all possible code paths:

1. _No `FILE_FLAG_NO_BUFFERING`_: __Untested__, will issue multiple syscalls, `TryPrepareScatterGatherBuffers` won't be even called. Subsequent items assume this flag is used.
2. _Buffers aligned to both disc sector and memory page boundaries, one memory page long, but whose total size does not fit in a 32-bit integer_: __Untested__, will issue multiple syscalls.
3. _Buffers aligned to both disc sector and memory page boundaries, and one memory page long_: __Untested__, the best case; will issue one syscall.
4. _Buffers aligned to both disc sector and memory page boundaries, and not one memory page long_: __Tested__, will issue multiple syscalls.
5. _Buffers aligned to disc sector boundaries, but not memory page boundaries_: __Mostly theoretical__ since there are no known disks with sectors larger than 4K, but it would issue multiple syscalls.
6. _Buffers not aligned to disc sector boundaries_: __Meaningless__, since we use `FILE_FLAG_NO_BUFFERING` and violate its rules, such test is _guaranteed to fail_.

Of these, testing (1) would better happen in a separate PR since it's applicable to both Windows and Unix. Testing (2) would require lots of memory and either way this and (1) are pre-existing code paths.

I will however modify my test to also test (3).